sql: make TestSchemaChangeRetryOnVersionChange retry on abort safe

vivekmenezes opened the pull request:
The internal transactions used by the schema change would occasionally
retry on abort, reevaluating the testing knob functions called
within the transactions producing incorrect results. This change makes
the test retry on abort safe.

fixes #13878

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13900)
<!-- Reviewable:end -->

vivekmenezes committed:
sql: make TestSchemaChangeRetryOnVersionChange retry on abort safe

The internal transactions used by the schema change would occasionally
retry on abort, reevaluating the testing knob functions called
within the transactions producing incorrect results. This change makes
the test retry on abort safe.

fixes #13878

andreimatei commented:
I've got some questions :)
1. Why exactly are backfill transactions being aborted? Maybe we can control the test so that they're not?
1. Is there a point in associating the backfill distsql flows with a txn? If not, I think we should use a different txn for resolving spans every time, instead of having a single txn spanning the whole execution of the backfill. https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/backfill.go#L640
1. The test in question in this PR, `TestSchemaChangeRetryOnVersionChange`, checks that a backfill is retried at a high level (`SchemaChanger.runBackfill()` is called again) when the version of a descriptor changes while we're in the middle of a backfill (I think...). 
a) I found it difficult to tell how that happens. I think it's because `backfillIndexes()` returns a `version mismatch` error. Can we test that directly, besides (or even instead of) counting how many times something gets called?
b) Is this test important? How can a table version change while a schema change is in progress? Only schema changes can increment the version, and schema changes require the schema change lease, right? If the table version can somehow be incremented while a backfill is in progress, is this the retry mechanism we want? Why exactly does a backfill need to be interrupted by the fact that the version changed?

 

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/schema_changer_test.go, line 1114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZVlaGZqbsin_uM3za:-KeZVlaGZqbsin_uM3zb:b5tjh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1114)):*
> ```Go
> 	var upTableVersion func()
> 	attempts := 0
> 	// This represents the number of backfill chunks that get reevaluated.
> ```

I'm sorry, but I can't tell what this test is doing. These 3 different counters make it hard to read. Maybe better names or comments would help.

---

*[pkg/sql/schema_changer_test.go, line 1118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZU6MK54g-TxofW_80:-KeZU6MK54g-TxofW_81:b-2wyg1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1118)):*
> ```Go
> 	var numReevaluated uint32
> 	var numBackfills, backfillAttempt uint32
> 	seenSpan := roachpb.Span{}
> ```

rename `seenSpan` to `lastSpan`?

---

*[pkg/sql/schema_changer_test.go, line 1134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZUwJilMDxI59Dt4mh:-KeZUwJilMDxI59Dt4mi:b86cv92) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1134)):*
> ```Go
> 				if seenSpan.Key != nil {
> 					// Keep track of the number of reevaluations.
> 					if b := atomic.LoadUint32(&numBackfills); b != backfillAttempt &&
> ```

I don't understand this condition... We're checking if the whole backfill restarted *and* whether we're processing a span lower than the last one? A comment about transaction retries around here somewhere would probably help.

---

*[pkg/sql/schema_changer_test.go, line 1152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZSk9JF4GA3DG3MpXy:-KeZSk9JF4GA3DG3MpXz:b-ajm2vv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1152)):*
> ```Go
> 		},
> 		DistSQL: &distsqlrun.TestingKnobs{
> 			RunBeforeBackfillChunk: func(sp roachpb.Span) error {
> ```

how come we need to duplicate these hooks (`SQLSchemaChanger` vs `DistSQL`)? Is it because only some of the backfills currently go through distsql? Please add a comment.
And can we extract a common function.

---

*[pkg/sql/schema_changer_test.go, line 1219 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZQyofN1EiBoE9V53e:-KeZQyofN1EiBoE9V53f:bm88dey) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1219)):*
> ```Go
> 	// the gateway node and therefore rely on the gateway node eventually
> 	// reevaluating the schema change on a version change.
> 	if reevaluated := atomic.SwapUint32(&numReevaluated, 0); reevaluated != 0 {
> ```

how come `numReevaluated` and friends need to be atomics? `addIndexSchemaChange` waits for the backfill to complete, so it shouldn't run concurrently with this code, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-:-KeZXJgqMj9LSdsvoLPF:b-u80teb)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
1. I'm not sure yet. I'll investigate and get back to you on that
2. we are using a txn only for the table lease. I think we can change that. Will send you another PR for that.
3. I think the main problem here is that we don't have a way to control two nodes not run the schema changer, since the
lease is not guaranteed to be held by one. I had added this test because when a table gets dropped in the middle of a schema change, the drop can get activated by another node, if it is able to somehow gain the lease. I think it's worth creating an issue to drop this test of figure out a way in which we can eliminate the need for this test in the future.

Thanks for the review. I haven't addressed all your comments yet in particular the one about why this transaction is aborting under stress. I'll get back to you on that.

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/schema_changer_test.go, line 1114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZVlaGZqbsin_uM3za:-KedDi6lghO4IpaaOcy1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1114)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm sorry, but I can't tell what this test is doing. These 3 different counters make it hard to read. Maybe better names or comments would help.
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 1118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZU6MK54g-TxofW_80:-KedDidSm-KNPJSxq6_x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1118)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

rename `seenSpan` to `lastSpan`?
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 1134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZUwJilMDxI59Dt4mh:-KedDlOgRTUoGPfVlFsx:b-dyxo7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1134)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't understand this condition... We're checking if the whole backfill restarted *and* whether we're processing a span lower than the last one? A comment about transaction retries around here somewhere would probably help.
</blockquote></details>

Added a comment and renamed the variables. Do let me know if . it's still hard to read.

---

*[pkg/sql/schema_changer_test.go, line 1152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZSk9JF4GA3DG3MpXy:-KedDtJ8HiNdgLT7aoxr:bxv70f7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1152)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how come we need to duplicate these hooks (`SQLSchemaChanger` vs `DistSQL`)? Is it because only some of the backfills currently go through distsql? Please add a comment.
And can we extract a common function.
</blockquote></details>

All the tests in this file repeat the code in both hooks because we haven't yet transitioned column backfills to distsql. I had originally thought of refactoring into a separate function, but then I figured the original copy is going to die soon.

---

*[pkg/sql/schema_changer_test.go, line 1219 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-KeZQyofN1EiBoE9V53e:-KedF9YcJVpkCGa4_htp:bp4xxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc24bd4764f8527007a484109e7275cab8b8304/pkg/sql/schema_changer_test.go#L1219)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how come `numReevaluated` and friends need to be atomics? `addIndexSchemaChange` waits for the backfill to complete, so it shouldn't run concurrently with this code, right?
</blockquote></details>

Hmmm, I'm note sure why it was made an atomic. Let me change that in a separate change. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13900#-:-KedFG9-tMkEO2VwD2T9:bsabgnw)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
closing in favor of #14131 . Turns out we were incrementing the version by more than 1 (v + 2) and as a result waiting on (v + 1) instead of (v). As a result the schema changer would sometimes race and pick up a lease on (v) resulting in more than one retry. This had nothing to do with transactions aborting and retrying.

vivekmenezes commented:
https://github.com/cockroachdb/cockroach/issues/14156 open because of the discussion in this PR
